src/main/java/com/google/devtools/build/lib/remote/RemoteRepoContentsCacheImpl.java

```diff
@@ -0,0 +1,311 @@
+  private static final String MARKER_FILE_PATH = ".recorded_inputs";
+  private static final String REPO_DIRECTORY_PATH = "repo_contents";
+
+  private static final Command COMMAND =
+      Command.newBuilder()
+          // A unique but nonsensical command that is valid on all platforms. It is never executed,
+          // but should pass all checks that an RE backend may apply to commands.
+          .addArguments(GUID.toString())
+          .addOutputPaths(MARKER_FILE_PATH)
+          .addOutputPaths(REPO_DIRECTORY_PATH)
+          .addOutputFiles(MARKER_FILE_PATH)
+          .addOutputDirectories(REPO_DIRECTORY_PATH)
+          .build();
+  private static final Directory INPUT_ROOT = Directory.getDefaultInstance();
+
+  private final CombinedCache cache;
+  private final String buildRequestId;
+  private final String commandId;
+  private final boolean acceptCached;
+  private final boolean uploadLocalResults;
+  private final DigestUtil digestUtil;
+  private final Action baseAction;
+
+  public RemoteRepoContentsCacheImpl(
+      CombinedCache cache,
+      String buildRequestId,
+      String commandId,
+      boolean acceptCached,
+      boolean uploadLocalResults) {
+    this.buildRequestId = buildRequestId;
+    this.commandId = commandId;
+    this.cache = cache;
+    this.acceptCached = acceptCached;
+    this.uploadLocalResults = uploadLocalResults;
+    this.digestUtil = cache.digestUtil;
+    this.baseAction =
+        Action.newBuilder()
+            .setCommandDigest(digestUtil.compute(COMMAND))
+            .setInputRootDigest(digestUtil.compute(INPUT_ROOT))
+            .build();
+  }
+
+  @Override
+  public void addToCache(
+      RepositoryName repoName,
+      Path fetchedRepoDir,
+      Path fetchedRepoMarkerFile,
+      String predeclaredInputHash,
+      ExtendedEventHandler reporter)
+      throws InterruptedException {
+    var context = buildContext(repoName);
+    if (!context.getWriteCachePolicy().allowRemoteCache()) {
+      return;
+    }
+    try {
+      if (FileSystemUtils.readLinesAsLatin1(fetchedRepoMarkerFile).stream()
+              .filter(line -> !line.isEmpty())
+              .count()
+          != 1) {
+        // This cache currently only supports marker files that contain nothing but the predeclared
+        // inputs hash. Repo rules with dependencies expressed only at runtime would require a
+        // two-stage cache lookup. Among the rules that are supported are http_archive and
+        // git_repository without patches.
+        return;
+      }
+    } catch (IOException e) {
+      reporter.handle(
+          Event.warn(
+              "Failed to read marker file repo %s, skipping: %s"
+                  .formatted(repoName, e.getMessage())));
+    }
+    var action = buildAction(predeclaredInputHash);
+    var actionKey = new ActionKey(digestUtil.compute(action));
+    var remotePathResolver =
+        new RemotePathResolver() {
```

Comment:
let's move this to a nested class elsewhere -- all the unsupported overridden methods are quite distracting